### PR TITLE
GH-100479: Add optional `blueprint` argument to `pathlib.PurePath`

### PR DESCRIPTION
**Pull request by barneygale**
Add optional *blueprint* argument to `pathlib.PurePath` and `Path`. This argument is supplied whenever a derivative path is created, such as from `PurePath.parent`. Subclasses may use this to pass information between paths.

* Issue: gh-100479

### CODE REVIEW

File: Lib/pathlib.py
Comment (merwok) [inline Lib/pathlib.py]: Why this change?

Comment (barneygale) [reply Lib/pathlib.py]: Technically not necessary for this patch, but it aligns `cwd()` and `absolute()` with `home()` and `expanduser()`, i.e. the classmethod calls through to the instance method, and not vice-versa.

This should give a small performance improvement in `path.absolute()` (which I suspect is used much more than `Path.cwd()`), because it doesn't need to construct _two_ path objects.

Comment (barneygale) [reply Lib/pathlib.py]: I could move this to another PR if you thought it appropriate?

**Comment by barneygale** [comment]
Hey @serhiy-storchaka, I believe you're pretty familiar with the pathlib internals. Penny for your initial impressions of this patch? It does cut out a key part of the initial pathlib design: the idea of skipping re-parsing and re-normalizing paths in case where we can be sure it's not necessary (e.g. in `path.parents`, `path.iterdir()`). I'd love to be able to convince you this is a good idea!

File: Lib/pathlib.py
Comment (merwok) [reply Lib/pathlib.py]: The previous implementation was straightforward and matched exactly what the docstring said.
Now it is more indirect and it’s not obvious that it does the exact same thing.

I feel that «aligning» is not a strong reason to change here (and I don’t understand the comment about helping performance of `absolute`).

Comment (barneygale) [reply Lib/pathlib.py]: I'll undo the change as I'm failing to easily justify it :-).

I'll log it as a separate (minor) performance bug with a fuller explanation.

**Comment by merwok** [comment]
Please remember no force-pushes for Python projects!

**Comment by barneygale** [comment]
Apologies. Too much Christmas wine.

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: FYI, I logged this as #100562

**Comment by barneygale** [comment]
This is now ready for review! My previous comment no longer applies, as we fixed the pathlib construction weirdness in #102789. This change should be pretty performance-neutral now.

File: Lib/pathlib.py
Comment (zmievsa) [inline Lib/pathlib.py]: This feels like a change in logic. Is it intentional? I.e. If len(parts) == 0 and not drv and not root -- we return self

File: Lib/pathlib.py
Comment (zmievsa) [inline Lib/pathlib.py]: This makes a lot of sense as a concept: `__new__` requires much more thought during subclassing.

However, are we certain that unpacking is the best option for performance? Have you tried benchmarking it?

**Review by zmievsa — APPROVED** [review]
After looking through the changes, this is an obvious approve from my side: the changes make a lot sense and we have been  waiting for them for a long time.

My biggest concern is that we're adding a breaking change to `_from_parsed_parts` which is not a big problem on its own because the interface is private, but I would still take a look around for some relatively popular libraries that might actually be dependent on that interface.

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: Yeah, it's intentional. The idea of this `if:` block is to `return self` is we're at the top level, as an optimization, because:

```python
Path('/').parent == Path('/')
Path('.').parent == Path('.')
```

Previously it was working for the first case but not the second.

I used the same logic in the re-implementation of `parents` below, so it made sense to copy it here.

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: Are you suggesting a signature like `makepath(self, args)`? It might be slightly faster but it's incongruous with the initialiser and `joinpath()`

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: To be clear, this method would still work if the `if:` block was taken out

File: Lib/pathlib.py
Comment (zmievsa) [reply Lib/pathlib.py]: Oooh right... I compared it to `_make_child_relpath` instead but comparing to a public interface is more fair. I guess you're right, consistency takes precedence.

**Comment by barneygale** [comment]
> My biggest concern is that we're adding a breaking change to `_from_parsed_parts` which is not a big problem on its own because the interface is private, but I would still take a look around for some relatively popular libraries that might actually be dependent on that interface.

Well, it's undocumented and underscore-prefixed. We've been breaking other things like that over the last few months, e.g. removing `Path._accessor`. It's hard to make progress without doing this.

File: Doc/library/pathlib.rst
Comment (barneygale) [inline Doc/library/pathlib.rst]: FYI @AlexWaygood, changing this to a tuple is probably the most questionable part of the patch. Would be grand to get your input on whether it's reasonable or not! Thanks

File: Lib/pathlib.py
Comment (AlexWaygood) [inline Lib/pathlib.py]: Hmm, wouldn't this be simpler as a classmethod?

```suggestion
    @classmethod
    def makepath(cls, *args):
        """Construct a new path object from any number of path-like objects.
        Subclasses may override this method to customize how new path objects
        are created from methods like `iterdir()`.
        """
        return cls(*args)
```

Comment (Gobot1234) [reply Lib/pathlib.py]: Making it a classmethod means that you can't share state between paths no?

File: Doc/library/pathlib.rst
Comment (AlexWaygood) [reply Doc/library/pathlib.rst]: Hmm, I'm pretty uncomfortable about doing this without emitting a deprecation warning, I'm afraid :(

It feels like quite a big backwards-compatibility break, given that you can do this on `main`:

```pycon
>>> from pathlib import Path
>>> x = Path("foo", "bar", "baz").parents
>>> x
<WindowsPath.parents>
>>> x[0]
WindowsPath('foo/bar')
>>> x[1]
WindowsPath('foo')
```

File: Lib/pathlib.py
Comment (AlexWaygood) [reply Lib/pathlib.py]: I see -- sorry, just getting back to pathlib stuff after a while away!

I guess I find the "makepath" name quite unintuitive in that case. The name to me implies that it's "just" an alternative constructor, which I'd expect to be a classmethod. Maybe something like "sproutpath" (feel free to bikeshed the name), which is clearer that it is _using the state of the current `Path` instance_ would be better?

File: Doc/library/pathlib.rst
Comment (AlexWaygood) [inline Doc/library/pathlib.rst]: Should be plural, I think, similar to the way we have `*args` rather than `*arg`:

```suggestion
.. method:: PurePath.makepath(*others)
```

**Review by AlexWaygood — COMMENTED** [review]
Sorry for taking so long to get to this!!

I think this needs a little more work for now. ~~I'm not comfortable with the change to the type of `.parents`; it feels like it could break currently working code that relies on being able to get a `Path` object back when indexing into the result of `.parents` attribute access.~~ (EDIT: I was wrong, this isn't a concern.) I'm also not sure "makepath" is the best name for this new method.

I'm also slightly concerned that this design might be a little fragile. The following situation seems fairly plausible:

1. Barney takes a break from open source for a few months
2. During Barney's break, a contributor files a PR to add `awesome_new_method` to pathlib. The method is so awesome, it's immediately clear to the core dev team that we desperately need the method.
3. Alex merges the PR, but because Barney's on an open-source break, there's nobody to remind us that the new method should really be calling `self.makepath(*args)` instead of `self.__class__(*args)`. Support for sharing state between pathlib subclasses is now broken :(

Is there a way we can guard against this?

**Comment by barneygale** [comment]
Thanks for taking a look!

> Sorry for taking so long to get to this!!
> 
> I think this needs a little more work for now. I'm not comfortable with the change to the type of `.parents`; it feels like it could break currently working code that relies on being able to get a `Path` object back when indexing into the result of `.parents` attribute access. 

That still works, unless I'm misunderstanding you

> I'm also not sure "makepath" is the best name for this new method.

Yeah, I'm not sure either. `zipfile.Path`'s corresponding method is called `_next()`. I chose `makepath()` due to its similarity to `joinpath()`, which already exists and has a similar purpose and signature.

> I'm also slightly concerned that this design might be a little fragile. The following situation seems fairly plausible:
> 
>     1. Barney takes a break from open source for a few months
> 
>     2. During Barney's break, a contributor files a PR to add `awesome_new_method` to pathlib. The method is so awesome, it's immediately clear to the core dev team that we desperately need the method.
> 
>     3. Alex merges the PR, but because Barney's on an open-source break, there's nobody to remind us that the new method should really be calling `self.makepath(*args)` instead of `self.__class__(*args)`. Support for sharing state between pathlib subclasses is now broken :(
> 
> 
> Is there a way we can guard against this?

Hm. I expect a user will spot that it doesn't work and log a bug. It doesn't have any wider consequences, I don't think.

File: Doc/library/pathlib.rst
Comment (barneygale) [reply Doc/library/pathlib.rst]: With this PR:

```python
>>> from pathlib import Path
>>> x = Path("foo", "bar", "baz").parents
>>> x
(PosixPath('foo/bar'), PosixPath('foo'), PosixPath('.'))
>>> x[0]
PosixPath('foo/bar')
>>> x[1]
PosixPath('foo')
```

Comment (AlexWaygood) [reply Doc/library/pathlib.rst]: Yep, you're quite right! I think I mistyped something in the REPL, and confused myself :)

Many apologies!

File: Lib/pathlib.py
Comment (AlexWaygood) [inline Lib/pathlib.py]: The parameter should have the same name you give it in the docs:

```suggestion
    def makepath(self, *pathsegments):
        """Construct a new path object from any number of path-like objects.
        Subclasses may override this method to customize how new path objects
        are created from methods like `iterdir()`.
        """
        return type(self)(*pathsegments)
```

File: Doc/library/pathlib.rst
Comment (barneygale) [reply Doc/library/pathlib.rst]: I've switched it to `*pathsegments`, matching the documentation for the initialiser. Is that any good?

File: Lib/pathlib.py
Comment (Gobot1234) [reply Lib/pathlib.py]: You still need the same for joinpath as well right?

File: Doc/library/pathlib.rst
Comment (AlexWaygood) [reply Doc/library/pathlib.rst]: it's good!

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: `with_segments()`, perhaps? Complements `with_name()` etc.

Comment (AlexWaygood) [reply Lib/pathlib.py]: `with_segments` sounds like it's kiiinda "the previous path's segements, but with some new ones added at the end".

What about "newpath"? I think that reinforces the fact that the returned object has completely different segments to the old one.

**Comment by AlexWaygood** [comment]
> > I'm also slightly concerned that this design might be a little fragile. The following situation seems fairly plausible:
> > ```
> > 1. Barney takes a break from open source for a few months
> > 
> > 2. During Barney's break, a contributor files a PR to add `awesome_new_method` to pathlib. The method is so awesome, it's immediately clear to the core dev team that we desperately need the method.
> > 
> > 3. Alex merges the PR, but because Barney's on an open-source break, there's nobody to remind us that the new method should really be calling `self.makepath(*args)` instead of `self.__class__(*args)`. Support for sharing state between pathlib subclasses is now broken :(
> > ```  
> >   
> > Is there a way we can guard against this?
> 
> Hm. I expect a user will spot that it doesn't work and log a bug. It doesn't have any wider consequences, I don't think.

Okay, I chatted with Brandt at the sprint, and he persuaded me that I was worrying too much about this hypothetical scenario :)

**Comment by barneygale** [comment]
FWIW, I am open to alternatives to this new method. There are things we can do like pass an opaque object around, or derive a new type to 'bind' the context. I went back and forth with these, and in the end went with this approach as it appears to be the simplest. But happy to look again if you have suggestions :)

**Comment by Gobot1234** [comment]
Also as a bit of a testimonial I really like the design and have the changes for something I locally implemented for this PR and it makes my code significantly nicer

File: Lib/pathlib.py
Comment (AlexWaygood) [inline Lib/pathlib.py]: The current implementation of `_PathParents` goes to great lengths to only create path objects "lazily" on request. That means that replacing `_PathParents` with a simple tuple creates quite a bad performance regression for this (unrealistic, artificial) benchmark.

On `main`:

```
>python -m timeit -s "from pathlib import Path; p = Path(*['foo' for _ in range(1_000)])" "p.parents[100]"
Running Debug|x64 interpreter...
2000 loops, best of 5: 118 usec per loop
```

With this PR:

```
>python -m timeit -s "from pathlib import Path; p = Path(*['foo' for _ in range(1_000)])" "p.parents[100]"
Running Debug|x64 interpreter...
5 loops, best of 5: 69.4 msec per loop
```

(These numbers are on a debug build, so take them with a pinch of salt, but I think it's highly likely a non-debug build would also show a significant regression on this benchmark.)

The benchmark above is, as I say, quite artificial, but I think it's reasonable to be concerned about this. `Path.parents` must remain a property (to maintain backwards compatibility), but you generally shouldn't have properties that do a significant amount of work and could potentially become performance bottlenecks -- those should be methods rather than properties, to make it explicit that "work is being done here". (Generally speaking, nobody "expects" the "`.` operator" to do a significant amount of work.) Also, the current code does seem to work quite hard to only create new path objects when absolutely necessary, which makes me worry that this could be quite performance-sensitive.

Comment (barneygale) [reply Lib/pathlib.py]: Fair point! I could restore the `_PathParents` class, but it would need to keep a reference to its superior path object. This is avoided at the moment:

https://github.com/python/cpython/blob/c9134fb228860bbb641794b1d165a95a55b5c669/Lib/pathlib.py#L215-L220

I don't really understand the comment as it doesn't look like we'd create a reference _cycle_. What do you think?

Comment (barneygale) [reply Lib/pathlib.py]: I've restored the `_PathParents` class. I think the comment about reference cycles is a mistake or a leftover from an early implementation.

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: Although this method is mostly called under-the-hood, it may also be useful when dealing with subclasses of pathlib classes. For example:

```python
from tarfile import TarFile, TarPath
readme_path = TarPath('README.txt', tarfile=TarFile('blah.tar.gz'))
license_path = readme_path.newpath('LICENSE.txt')
```

I mention this just in case it helps us with the naming question. In the above example, `readme_path.newpath()` is a sort of factory for path objects sharing the same underlying `TarFile` object. And I suppose factories _make_ things. But I don't mind `newpath()` :)

File: Lib/pathlib.py
Comment (AlexWaygood) [reply Lib/pathlib.py]: Yeah -- there doesn't appear to be any kind of reference cycle here, since the `.parents` property returns a new `_PathParents` object each time! So I think we're good:

https://github.com/python/cpython/blob/ae4845441281e9b6cf2623388df28587afa239c2/Lib/pathlib.py#L646-L649

Interestingly, one possible optimisation here _could_ be to make `.parents` a cached property. But that might not gain us very much (it looks to me like `_PathParents` objects should be pretty cheap to construct? But I haven't measured it), and if we did that, we really _would_ have to worry about a reference cycle.

Maybe it would be worth, though, adding a comment to the `.parents` property warning us about the possibility of a reference cycle if we were to cache the result of the property?

File: Lib/pathlib.py
Comment (AlexWaygood) [reply Lib/pathlib.py]: Of the options floated so far, I like the "newpath" name the best. But it's your module now, afterall, so I don't feel like I should have the final call here :D

You could maybe start a thread on Discord or Discuss if you'd like the opinion of more core devs?

**Review by AlexWaygood — APPROVED** [review]
The implementation and the docs here look great. Thanks for addressing my concerns, and for your patience :)

My only remaining reservation is about the name of the method itself. I think I would still prefer "newpath" -- but if you'd like more opinions, you could always start a thread on Discourse or Discord :)

**Comment by barneygale** [comment]
Thanks so much for the thorough review! I'll open a thread about the method name shortly.

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: Thread created here! https://discuss.python.org/t/bikeshedding-opportunity-help-name-a-pathlib-method/26186

File: Doc/library/pathlib.rst
Comment (AlexWaygood) [inline Doc/library/pathlib.rst]:
```suggestion
      The classes provided in this module ignore the *template* argument.
      It is there purely as a hook for user-defined subclasses.
```

File: Misc/NEWS.d/next/Library/2023-04-03-22-02-35.gh-issue-100479.kNBjQm.rst
Comment (AlexWaygood) [inline Misc/NEWS.d/next/Library/2023-04-03-22-02-35.gh-issue-100479.kNBjQm.rst]: You should totally take credit here :)

```suggestion
to pass information to derivative paths. Patch by Barney Gale.
```

**Review by AlexWaygood — COMMENTED** [review]
Wow, using the `template` argument instead of adding the `makepath` method makes it so much easier for me to understand what's going on! This is a great improvement!

**Comment by barneygale** [comment]
Yeah, I was missing the wood for the trees there. Massive improvement.

**Review by AlexWaygood — APPROVED** [review]
LGTM. I guess the only remaining question is whether "template" is the best name for this new parameter. I'm fine with "template", but you might want to head back to Discourse if you want some more bikeshedding :)

File: Doc/library/pathlib.rst
Comment (AlexWaygood) [inline Doc/library/pathlib.rst]: Should we add a test to make sure diamond inheritance works?

Comment (barneygale) [reply Doc/library/pathlib.rst]: I'm not sure. There's only one place in `pathlib.py` where we call `super()`, and that only exists because we need to raise a deprecation warning when additional arguments are supplied to `pathlib.Path()`. The `Path.__init__()` method will be removed in 3.14, at which point it will be impossible for the test to fail.

File: Doc/library/pathlib.rst
Comment (barneygale) [inline Doc/library/pathlib.rst]: I've written this out in full for clarity's sake, but it could be:

```suggestion
              self.session_id = template.session_id if template else session_id
```

Comment (zmievsa) [reply Doc/library/pathlib.rst]: I personally prefer the full version because the modern python coverage tools cannot handle ternary expressions.

File: Doc/library/pathlib.rst
Comment (barneygale) [reply Doc/library/pathlib.rst]: Perhaps a hidden `.. doctest::` block would be best?

File: Doc/library/pathlib.rst
Comment (AA-Turner) [inline Doc/library/pathlib.rst]: Does it make sense to specify that ``template: Self | None`` here? I.e. that if template is not None, it will be an instance of the current (user-defined) class.

A

**Comment by barneygale** [comment]
I've created a branch/PR that uses a `__newpath__()` method, for comparison's sake:

- #103975

**Comment by barneygale** [comment]
Closing in favour of #103975.

**Pull request closed without merging**
